### ISSUE TITLE
[Bug] min()/max() on empty series returns large/small value instead of nan

### ISSUE THREAD

**Issue by ayushdg** [gpuopenanalytics/pygdf#195]
## Reporting a bug

- [X] I am using the latest version of PyGDF from conda ~~or built from `master`~~.
- [X] I have included the following environment details:
      Linux Distro: Ubuntu 16.04, Linux Kernel: Linux 4.15.0-29-generic x86_64 , GPU Model: TITAN V
- [X] I have included the following version information for:
      Arrow: 0.7.1, CUDA: 9.2.148, Numpy: 1.14.5, Pandas: 0.23.3, Python: 3.6.6
- [X] I have included below a minimal working reproducer (if you are unsure how
 to write one see http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

Calling `min()` or `max()` on an empty series returns largest/smallest value for the dtype of the series respectively. 
```
import pygdf
from pygdf import Series

gseries = Series([])
print(gseries.min())
print(gseries.max())
```

Output:
```
1.7976931348623157e+308
-1.7976931348623157e+308
```

Pandas returns `nan` for similar cases.

----
Workaround: Check if the series is empty before calling `min()`/`max()`

**Comment by pearu** [comment]
numpy raises a `ValueError: zero-size array to reduction operation minimum which has no identity` when calling `min` method of a empty array.
I suggest numpy behavior as it might be difficult to track down the origin of `nan` in the result of some algorithm.

**Comment by pearu** [comment]
[ANSI SQL appears to return `NULL`](https://dba.stackexchange.com/questions/25435/why-does-ansi-sql-define-sumno-rows-as-null) when `min` and other reduction functions are applied to empty table.

`NULL` and `nan` have different meanings. Still thinking that `nan` would be inappropriate in this situation but I guess this is debatable view.

**Comment by mrocklin** [comment]
We might want to return a NULL value.  I agree that we don't want to return
NaN.  I think that this is a case where deviating from Pandas behavior is
probably nice.  I suspect that Pandas also wanted to return NULL, but
couldn't because they don't have access to the value.

**Comment by kkraus14** [comment]
Yes, let's return NULL.

**Comment by pearu** [comment]
@mrocklin @kkraus14 returning NULL is not really an option because this is Python..

Consider, for example, the `min` call that will lead to the following calling sequence:
```
Series.min
NumericalColumn.min                 # python/cudf/dataframe/numerical.py
cpp_reduce.apply_reduce('min', ...) # python/cudf/bindings/reduce.pyx
gdf_min                             # cpp/src/reductions/reductions.cu
...
gpu_reduction_op
```
where `gpu_reduction_op` has the following implementation:
```
    T agg = identity;
    for (int base=blkid * blksz; base<size; base+=step) {
        int i = base + tid;
        T loaded = identity;
        if (i < size && gdf_is_valid(mask, i))
            loaded = loader(data, i);
        T temp = BlockReduce(temp_storage).Reduce(loaded, functor);
        agg = functor(agg, temp);
    }
```
Clearly, when the input is empty or has no valid data, the result will be`identity`. In the case of `min`, the value of `identity` is `std::numeric_limits<T>::max()`. Returning NULL would not be feasible even in C++ level because the result can be only a value.

I think the algorithm implemented in `gpu_reduction_op` is appropriate for reduction operations such as `sum` or `product` (that return `0` or `1` for empty data, respectively) but not for `min`, `max`, `mean`, and other such operations that do not have reasonable default values for empty sequences.

Instead of the current algorithm for `min`, for instance, (pseudo-code follows):
```
r = <largest value>
for d in valid_data:
    if d < r:
        r = d
``` 
I would propose the following algorithm (pseudo-code follows):
```
r = None
for d in valid_data:
    if r is None or d < r:
        r = d
if r is None:
    raise ValueError("reduction result undefined on non-valid or empty data")
```
for both Python and C++.

What do you think?

**Comment by mrocklin** [comment]
I recommend that we look to see what Pandas Int64 dtype (in the release
candidate) returns.  I wonder how they handle scalar NULLs.  cc
@TomAugspurger

**Comment by kkraus14** [comment]
> I recommend that we look to see what Pandas Int64 dtype (in the release candidate) returns. I wonder how they handle scalar NULLs.

It returns a Python `math.nan` object which is a float and non-numpy type.

**Comment by pearu** [comment]
Here are some relevant discussions from pandas:
  https://github.com/pandas-dev/pandas/issues/18200
in particular,
  https://github.com/pandas-dev/pandas/issues/18200#issuecomment-343747671
summarizes the dilemma well.

Recall the following `min([])` behaviors:
1. Python, numpy - raise ValueError exception
2. pandas.Series - returns nan (even for integer types)
3. cudf.Series - returns 1.8e308
4. SQL engines - return NULL (if these follow ANSI SQL standard, not always the case though)

We need a pick to resolve this issue.

**Comment by mrocklin** [comment]
To be clear when I say Int64 I mean the new, NULL-enabled integer dtypes in
Pandas

**Comment by kkraus14** [comment]
> To be clear when I say Int64 I mean the new, NULL-enabled integer dtypes in Pandas

The problem is the `max()` and `min()` operators return a Scalar, not a Series/Array, so there's no extension array to indicate a null/nan for an integer.

**Comment by TomAugspurger** [comment]
Pandas would probably return NULL if we had that option.

Our new nullable integer dtype returns NaN

```python
In [23]: pd.Series([], dtype=pd.Int32Dtype()).min()
Out[23]: nan
```

though I assume that's just for compatibility with the float version.

**Comment by pearu** [comment]
Pandas `min` could made a choice of returning `nan` (float value) on empty integer series because Python allows it. (Personally, I disagree with the choice for many reasons...)

cuDF should presumably be compatible to pandas. However, cuDF compute kernels are implemented in C++ which would not allow the choice made in pandas. If cuDF is required to be pandas compatible, we could implement that in Python layer.

**Comment by harrism** [comment]
I think it is reasonable for libcudf (the C++ part) to return a CUDF error such as `CUDF_DATASET_EMPTY` for empty input. Sure, it may be reasonable to return the identity (`0`) for `sum()` (zero of anything is zero), but is it reasonable to return ONE for a product of zero numbers (zero ones are ZERO, not ONE)? And as explained above, min/max of zero values also has no reasonable identity, so should probably return an error. The question is: do we return `0` for sum() and `CUDF_DATASET_EMPTY` for all other operators, or return the error for all operators?

The Python API can choose to match Pandas and either detect empty input and return what Pandas returns, or call libcudf, detect the error return value and then return what Pandas returns.

**Comment by pearu** [comment]
Yes, it is reasonable to return ONE for a product of zero numbers. All CAS do that.
See https://en.wikipedia.org/wiki/Empty_product  for reasoning.

To sum up, the reduction results of the empty sequences ought to be:
1. sum -> 0
2. product -> 1
3. min, max, mean, std, var, etc 
    -> in C++, return CUDF_DATASET_EMPTY error
    -> in Python, do what pandas does. If deviating from pandas is ok, then raise ValueError.

**Comment by pearu** [comment]
Surprisingly, `pandas.Series([]).product()` returns `nan`.

**Comment by TomAugspurger** [comment]
> Surprisingly, pandas.Series([]).product() returns nan.


```python
In [15]: pd.Series([]).prod()
Out[15]: 1.0
```

Can you check your pandas version? This changed somewhat recently (last year).

I would pretty strongly recommend following pandas for 

1. the sum of an empty series (0)
2. the product of an empty series (1)

We went through a long, painful process [last year](https://mail.python.org/pipermail/pandas-dev/2017-November/000657.html) clarifying / fixing this. People have strong opinions on it. And this matches many other libraries, most notably NumPy,

```python
In [20]: np.nansum([])
Out[20]: 0.0

In [21]: np.nanprod([])
Out[21]: 1.0
```


Pandas includes the `min_count` parameter, for users who want the sum / prod of an empty series to be NaN.

```python
In [16]: pd.Series([]).prod(min_count=1)
Out[16]: nan
```

if there aren't at least `min_count` non-NA observations, then the result is NA (nan).

---

min / max don't really have similar units (we didn't view the min / max integers for the dtype to be comparable), so those continue to return NaN.

**Comment by pearu** [comment]
@TomAugspurger thanks for pointing out the `pd.Series([]).prod()->1` feature in the latest pandas versions. I am happy that you ended up with this solution.

I am using pandas version 0.20.3 according to cudf/conda/environments/dev_py35.yml where empty product returns `nan`.

**Comment by kkraus14** [comment]
I'm perfectly happy making the Pandas-compatibility in the Python layer and the libcudf portions having more logical behavior.

**Comment by pearu** [comment]
The general aim of cudf is provide a pandas-like API for GPU accelerated workflow.
I think this gives a general guideline for making decisions that we were facing in this issue.

However, as seen above, the history of pandas behavior may change and this will influences also cudf code. Currently, [cudf supports only pandas version 0.20.3](https://github.com/rapidsai/cudf#customizing-the-build) which brings up an interesting problem: should we implement the pandas mistakes made in past, or should we skips these and aim at supporting the  recent versions of pandas behavior?

**Comment by kkraus14** [comment]
> The general aim of cudf is provide a pandas-like API for GPU accelerated workflow.
> I think this gives a general guideline for making decisions that we were facing in this issue.
> 
> However, as seen above, the history of pandas behavior may change and this will influences also cudf code. Currently, [cudf supports only pandas version 0.20.3](https://github.com/rapidsai/cudf#customizing-the-build) which brings up an interesting problem: should we implement the pandas mistakes made in past, or should we skips these and aim at supporting the recent versions of pandas behavior?

We will be bumping support to the latest Pandas version in #668, so lets aim for newest behavior.
